### PR TITLE
Add backpressure graph and color pipeline nodes

### PR DESCRIPTION
**Pull request by jbeisen**
Add a graph in the OperatorDetail component that displays the backpressure metric.

Color the operator nodes in the pipeline graph according to the maximum backpressure of all its subtasks.

Resolves: #75

---

image

### CODE REVIEW

File: arroyo-console/src/routes/pipelines/JobDetail.tsx [hunk 1]
```diff
@@ -399,14 +419,26 @@ function OperatorDetail({
           };
         });
       });
+
+      backpressureData = Object.entries(node_metrics.subtasks).map(kv => {
+        return kv[1].backpressure.map(m => {
+          return {
+            label: kv[0],
+            x: new Date(Number(m.time) / 1000),
+            y: m.value,
```
Comment (jbeisen) [inline arroyo-console/src/routes/pipelines/JobDetail.tsx:428]: Note that I didn't add the randomness that the other metrics have. @mwylde what's up with that?

Comment (mwylde) [reply arroyo-console/src/routes/pipelines/JobDetail.tsx:428]: It adds a bit of jitter that helps separate out the substasks on the graph

File: arroyo-console/src/routes/pipelines/JobDetail.tsx [hunk 2]
```diff
@@ -422,6 +454,13 @@ function OperatorDetail({
           Events TX
           <TimeSeriesGraph data={msgSentData} timeWindowMs={5 * 60 * 1000} />
         </Box>
+
+        {backpressureData?.some(m => m.length) && (
```
Comment (mwylde) [inline arroyo-console/src/routes/pipelines/JobDetail.tsx:458]: I'm not sure it makes sense to have this as a graph -- generally this doesn't vary over time much (at not on the 5 minute timescale of these graphs).

File: arroyo-console/src/routes/pipelines/JobGraph.tsx [hunk 3]
```diff
@@ -40,16 +58,32 @@ function PipelineGraphNode({
...
   activeOperator?: string;
 }) {
   const nodeTypes = useMemo(() => ({ pipelineNode: PipelineGraphNode }), []);
 
   const nodes = graph.nodes.map(node => {
+    let operatorBackpressure = 0;
+    if (metrics) {
+      // reduce the subtasks backpressure to an operator-level backpressure
+      const nodeMetrics = metrics.metrics[node.nodeId];
+      const backPressureMetrics = Object.entries(nodeMetrics.subtasks).map(kv => {
```
Comment (mwylde) [inline arroyo-console/src/routes/pipelines/JobGraph.tsx:77]: I think this is causing errors before metrics are available as the pipeline is starting up:
> TypeError: Cannot read properties of undefined (reading 'subtasks')

File: arroyo-console/src/routes/pipelines/JobGraph.tsx [hunk 4]
```diff
@@ -9,13 +9,31 @@ import {
   JobMetricsResp,
 } from "../../gen/api_pb";
 
+
+function getNodeColor(backpressure: number): string {
+  // backPressure should be a value between 0 and 100
+
+  const thresholds = [33, 66];
```
Comment (mwylde) [inline arroyo-console/src/routes/pipelines/JobGraph.tsx:16]: Something isn't working quite right here -- even when the pipeline is fully backpressured (i.e., the tx queues are at 0 capacity) I don't see any colors

Comment (mwylde) [reply arroyo-console/src/routes/pipelines/JobGraph.tsx:16]: BTW, if you want to test backpressure, the easiest way to simulate it is to add a sleep in an operator:

```
tokio::time::sleep(Duration::from_secs(1).await
```

File: arroyo-console/src/routes/pipelines/JobDetail.tsx (hunk 2 above)
Comment (jbeisen) [reply arroyo-console/src/routes/pipelines/JobDetail.tsx:458]: Fair, maybe we should just make it a badge to indicate the current backpressure status.

File: arroyo-console/src/routes/pipelines/JobGraph.tsx (hunk 4 above)
Comment (jbeisen) [reply arroyo-console/src/routes/pipelines/JobGraph.tsx]: This is fixed now. I changed the logic for getting the current backpressure.

File: arroyo-console/src/routes/pipelines/JobDetail.tsx (hunk 1 above)
Comment (jbeisen) [reply arroyo-console/src/routes/pipelines/JobDetail.tsx]: I've removed the graph.

File: arroyo-console/src/routes/pipelines/JobGraph.tsx (hunk 3 above)
Comment (jbeisen) [reply arroyo-console/src/routes/pipelines/JobGraph.tsx]: This is fixed now.

File: arroyo-console/src/routes/pipelines/JobDetail.tsx (hunk 2 above)
Comment (jbeisen) [reply arroyo-console/src/routes/pipelines/JobDetail.tsx]: I removed the graph and added a badge instead.

**Review by mwylde — APPROVED** [review]

**Pull request merged**